RavenDB-24020 Allow to parallelize HNSW work

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-24020 

### Additional description

This PR allows us to build HNSW graph in parallel, as well as restructure how we load the data for the graph to allow us to handle batching / pre-fetching in a better manner.

Indexing of 224,000 vectors dropped from 6 minutes to just over 2. 
Indexing of millions of vectors dropped to < 3 hours.

### Type of change
- [x] Optimization

### How risky is the change?

- [x] Moderate 

### Backward compatibility

- [x] Non breaking change

### Is it platform specific issue?

- [x] No

### Documentation update

- [x] No documentation update is needed 

### Testing by Contributor

- [x] It has been verified by manual testing

Benchmarks

- [x] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [x] No

### UI work

- [x] No UI work is needed

maciejaszyk commented on src/Voron/Data/Graphs/Hnsw.Parallel.cs:
```diff
@@ -0,0 +1,426 @@
+                        }
+                        finally
+                        {
+                            if (activeTasksCount == ++_completed)
```
`++_completed` is this safe? It can be called simultaneously by task, right?

ayende replied:
Yes, that is safe - we are doing some funky stuff here with the threading
Everything here is _single_ threaded, except the distance computation, which runs separately and in bulk.

maciejaszyk commented on src/Voron/Data/Graphs/Hnsw.cs:
```diff
@@ -417,6 +423,47 @@ private void LoadNodeIndexes(ref NativeList<long> nodeIds, ref NativeList<int> i
+            indexes.EnsureCapacity(nodeIds.Count);
+            for (int i = 0; i < nodeIds.Count; i++)
+            {
+                if (_nodeIdToIdx.TryGetValue(nodeIds[i], out var index))
```
We can use CollectionMarshall to retrieve the index

ayende replied:
No longer relevant

ppekrol reviewed (approved):
Thinking out loud, are there any backward compatibility concerns? When first part was generated via previous impl and new items will be handled by this impl

ppekrol commented:
@Lwiel @redknightlois please review
@ayende can you rebase?

redknightlois commented:
> Thinking out loud, are there any backward compatibility concerns? When first part was generated via previous impl and new items will be handled by this impl

There should be, no.

redknightlois commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -634,6 +634,13 @@ public virtual bool RunInMemory
         public int MaxNumberOfThreadsForLocalEmbeddingsGeneration { get; set; }
         
+        [Description("Expert: The maximum number of concurrent batches for HNSW distnace computation acceleration.")]
+        [DefaultValue(512)]
```
Typo 'distance'

redknightlois commented on src/Voron/Data/Graphs/Hnsw.Parallel.cs:
```diff
@@ -0,0 +1,582 @@
+                await SearchNearestAcrossLevelsAsync(insertedVector, currentMaxLevel);
+                for (int level = nodeRandomLevel; level >= 0; level--)
+                {
+                    int startingPointIndex = _nearestIndexes[level];
```
This algorithm could use some inline comments explaining what it does and the tradeoffs involved.

ayende commented:
There shouldn't be any backward compatibility concerns, no.

ayende committed:
RavenDB-24020 - Changing HNSW insert to parallel mode.

It is now possible to interleave the IO work of multiple nodes in the graph at the same time that we are doing distance computation in the thread pool. We do that by interleaving (using yield) multiple "concurrent" node placement routines that
share the same runner. The runner runs each of them in turn and amortize all access to the disk (to load vectors) and then execute the actual distance computation on the thread pool. The runner is in charge of this whole orchestration, with the
idea being that we can run both computation and disk work at the same time.
This is critical because if we have a graph to insert that is larger than memory, we may be stall by swapping for basically forever without this approach.

- Because we are now doing multiple in-flight nodes, we add the previous N (number of edges) nodes as presumptive edges to the current node, to "kick start" the graph wiring process. This ends up

- Using the .NET default thread pool because all attempts (blocking collection, channels, dedicated threads, etc) has ended up significantly worse in terms of performance.
  The design means that we have a max of N batches of work to do (where N is likely small and fixed) and those tend to be very short lived. So even under load, it isn't expected to fill the thread pool and cause a train.

- Add max batch count to allow user to control the concurrency level
- Adding debug endpoint to expose the internal structure of HNSW index
- Fixing generation of debug output for HNSW
- Allow to generate debug output on machines without chrome

ayende commented:
Rebased everything, as far as I'm concerned, ready for review.

ayende commented:
Current results, running the `vector.benchmark` suite:

Before:

```
00:05:04.6600788
8 - 8 = 100.00% matches
387 - 408 = 94.85% matches
781 - 808 = 96.66% matches
1158 - 1208 = 95.86% matches
1550 - 1608 = 96.39% matches
1937 - 2008 = 96.46% matches
2322 - 2408 = 96.43% matches
2695 - 2808 = 95.98% matches
```

After:
```
00:02:56.2510365
8 - 8 = 100.00% matches
384 - 408 = 94.12% matches
772 - 808 = 95.54% matches
1143 - 1208 = 94.62% matches
1532 - 1608 = 95.27% matches
1926 - 2008 = 95.92% matches
2313 - 2408 = 96.05% matches
2702 - 2808 = 96.23% matches
```

In other words:

```
Speed: After solution is 43.6% faster (5:04.66 to 2:56.25).

Accuracy Difference:

8: 0% (100% vs 100%)
408: -0.73% (94.85% vs 94.12%)
808: -1.12% (96.66% vs 95.54%)
1208: -1.24% (95.86% vs 94.62%)
1608: -1.12% (96.39% vs 95.27%)
2008: -0.54% (96.46% vs 95.92%)
2408: -0.38% (96.43% vs 96.05%)
2808: +0.25% (95.98% vs 96.23%)
```

Having said that, pretty sure that I missed something in the rebase, it is supposed to be _more_ accurate after all of this.

ayende committed:
RavenDB-24020 - Only consider the _inflight_ requests, not just those that were inserted just before this one. Turns out that it creates a better graph (because if we already added an item, we see it in the graph and don't need to also add it, I think)

ayende commented:
Here are more changes, which brings both better perf and more accuracy. The change is to consider _only_ the in-flight, which means that if we already added an item, it will be loaded only from the graph.
End result is that this matter most to those that needs to load items from the disk, so likely cause us to consider more paths in the graph that we wouldn't otherwise. 

```
00:02:50.3396845
8 - 8 = 100.00% matches
391 - 408 = 95.83% matches
787 - 808 = 97.40% matches
1154 - 1208 = 95.53% matches
1543 - 1608 = 95.96% matches
1933 - 2008 = 96.26% matches
2300 - 2408 = 95.51% matches
2686 - 2808 = 95.66% matches
```

```
Speed: After solution is 44.1% faster (5:04.66 to 2:50.34).

Accuracy Difference:

8: 0% (100% vs 100%)
408: +0.98% (94.85% vs 95.83%)
808: +0.74% (96.66% vs 97.40%)
1208: -0.33% (95.86% vs 95.53%)
1608: -0.43% (96.39% vs 95.96%)
2008: -0.20% (96.46% vs 96.26%)
2408: -0.92% (96.43% vs 95.51%)
2808: -0.32% (95.98% vs 95.66%)
```

ayende committed:
RavenDB-24020 - Properly handle small vectors preloading

maciejaszyk commented on src/Voron/Data/Graphs/Hnsw.Parallel.cs:
```diff
@@ -0,0 +1,835 @@
+    {
+        private int _nextNodeIndex;
+
+        public int MaxConcurrentBatches = 512;
```
Missed setting this from RavenDB configuration.

ayende replied:
It's here: 
https://github.com/ayende/ravendb/blob/ad94ccac29a0f863e40453fd49048b77415740b4/src/Corax/Indexing/IndexWriter.cs#L938

maciejaszyk replied:
That commit is not a part of this PR. Does the other branch contain any other changes that should be included in this PR?

ayende replied:
Thanks for noticing, fix.

ayende committed:
RavenDB-24020 - wiring MaximumConcurrentBatchesForHnswAcceleration
